### PR TITLE
Distributed nd sort for cpu/omp

### PR DESCRIPTION
**Pull request by mfoerste4**
Added CPU/OMP version for distributed sample sort
* supports multiple dimensions
* utilizes cpu collective communicator
** NOTE: all communications are global (w.r.t. task). Communication in sub-groups are implemented via partly empty all2allv
* OMP/CPU shares the implementation, only difference is the execution policy used by thrust which is utilized for most cost-intensive ops

The algorithmic structure is taken from the GPU implementation.

### CODE REVIEW

File: src/cunumeric/sort/sort_cpu.inl [hunk 1]
```diff
@@ -0,0 +1,1027 @@
...
+        // in case the storage *is* unbound -- the result will be appended to output_array
+        if (volume > 0 && !is_unbound_1d_storage) {
+          if (argsort) {
+            auto output = output_array.write_accessor<int64_t, DIM>(rect);
+            assert(output.accessor.is_dense_row_major(rect));
+            output_ptr = static_cast<void*>(output.ptr(rect.lo));
+          } else {
+            auto output = output_array.write_accessor<VAL, DIM>(rect);
+            assert(output.accessor.is_dense_row_major(rect));
+            output_ptr = static_cast<void*>(output.ptr(rect.lo));
```
Comment (ipdemes) [inline src/cunumeric/sort/sort_cpu.inl:993]: lines 988-989 and 992-993 are the same, I think you can move them out of the if/else

File: src/cunumeric/sort/sort_cpu.inl [hunk 2]
```diff
@@ -0,0 +1,1027 @@
...
+using namespace legate;
+
+// sorts inptr in-place, if argptr not nullptr it returns sort indices
+template <typename VAL, typename DerivedPolicy>
+void thrust_local_sort_inplace(VAL* inptr,
+                               int64_t* argptr,
+                               const size_t volume,
+                               const size_t sort_dim_size,
+                               const bool stable_argsort,
+                               const thrust::detail::execution_policy_base<DerivedPolicy>& exec)
```
Comment (magnatelee) [inline src/cunumeric/sort/sort_cpu.inl:46]: you could make the whole type a template parameter instead (i.e., `const DerivedPolicy& exec` would work too), as detail types like this are subject to change.

**Comment by magnatelee** [comment]
@mfoerste4 I didn't look at the algorithms that closely this time, as I know they were translated from the original cuda code that I spent quite some time on reading. one high-level comment though is that you may want to split up the `sample_sort_nd` function into smaller functions just for code maintenance and readability, but it's not mandatory for the merge.

File: src/cunumeric/sort/sort.h [hunk 3]
```diff
@@ -18,8 +18,13 @@
 
 #include "cunumeric/cunumeric.h"
 
+#include <thrust/sort.h>
+
 namespace cunumeric {
 
+using namespace Legion;
```
[the remark this answers is not in the document]
Comment (mfoerste4) [reply src/cunumeric/sort/sort.h:25]: Removed

File: src/cunumeric/sort/sort_cpu.inl (hunk 2 above)
Comment (mfoerste4) [reply src/cunumeric/sort/sort_cpu.inl:46]: Good hint, I was not aware of this.

**Comment by mfoerste4** [comment]
> @mfoerste4 I didn't look at the algorithms that closely this time, as I know they were translated from the original cuda code that I spent quite some time on reading. one high-level comment though is that you may want to split up the `sample_sort_nd` function into smaller functions just for code maintenance and readability, but it's not mandatory for the merge.

@magnatelee , thanks for the comments. I applied your review suggestions. You are right that the main control routine is quite large. I might give it another cleanup/restructuring on the next occasion.

**Pull request merged**